front_end/webserver.py

```diff
@@ -300,72 +339,115 @@ def get(self, course, assignment, problem):
             else:
-                if has_duplicate_title(get_problems(course, assignment), problem, problem_basics["title"]):
+                if content.has_duplicate_title(content.get_problems(course, assignment), problem, self.get_body_argument("title").strip()):
                     result = "Error: A problem with that title already exists in this assignment."
                 else:
-                    for data_url in set(problem_details["data_urls"].split("\n")):
-                        data_url = data_url.strip()
-                        if data_url != "":
-                            contents, content_type, extension = download_file(data_url)
-                            file_name = create_md5_hash(data_url) + extension
-                            write_data_file(contents, file_name)
-                            problem_details["data_urls_info"].append([data_url, file_name, content_type])
-
-                    expected_output, error_occurred = exec_code(settings_dict["back_ends"][problem_details["back_end"]], problem_details["answer_code"], problem_basics, problem_details)
+                    details_dict = {"instructions": self.get_body_argument("instructions").strip().replace("\r", ""), "back_end": self.get_body_argument("back_end"), "output_type": self.get_body_argument("output_type"), "answer_code": self.get_body_argument("answer_code").strip().replace("\r", ""), "answer_description": self.get_body_argument("answer_description").strip().replace("\r", ""), "test_code": self.get_body_argument("test_code").strip().replace("\r", ""), "credit": self.get_body_argument("credit").strip().replace("\r", ""), "data_url": self.get_body_argument("data_url").strip().replace("\r", ""), "show_expected": self.get_body_argument("show_expected") == "Yes", "show_test_code": self.get_body_argument("show_test_code") == "Yes", "show_answer": self.get_body_argument("show_answer") == "Yes", "expected_output": ""}
 
+                    expected_output, error_occurred = exec_code(settings_dict["back_ends"][problem_details["back_end"]], self.get_body_argument("answer_code").strip().replace("\r", ""), problem_basics, details_dict)
+                    
                     if error_occurred:
                         result = "Error: " + expected_output
                     else:
                         problem_details["expected_output"] = expected_output
 
-                        save_problem(problem_basics, problem_details)
-                        problem_basics = get_problem_basics(course, assignment, problem)
-                        problem_details = get_problem_details(course, assignment, problem, parse_data_urls=True)
+                        if new_problem:
+                            content.save_problem(course, assignment, problem_basics, problem_details)
+                        else:
+                            if problem_basics["title"] != self.get_body_argument("title").strip():
```

Comment:
Please create local variables for title and introduction and visible, etc. so you don't need to repeatedly pull the body arguments.